### PR TITLE
*: prepare for raftstore-v2's split

### PR DESCRIPTION
**Pull request by SpadeA-Tang**
Signed-off-by: SpadeA-Tang <u6748471@anu.edu.au>

### What is changed and how it works?

Issue Number: Ref #12842

What's Changed:

This PR add a new API `create_checkpoint` which uses `create_checkpoint` of the rocksdb.
It also refactors a small part of v1's split in order to reduce code duplication when implementing v2.

### Check List 

Tests 

- Unit test
- Integration test

### Release note 

```release-note
None
```

### CODE REVIEW

File: components/raftstore/src/store/fsm/apply.rs [hunk 1]
```diff
@@ -1899,6 +1899,42 @@ mod confchange_cmd_metric {
     }
 }
 
+// Validate the request and the split keys
+pub fn validate_and_get_split_keys(
```
Comment (BusyJay) [inline components/raftstore/src/store/fsm/apply.rs:1903]:
```suggestion
pub fn extract_split_keys(
```

File: components/raftstore/src/store/fsm/apply.rs [hunk 2]
```diff
@@ -1899,6 +1899,42 @@ mod confchange_cmd_metric {
...
+            return Err(box_err!(
+                "invalid new peer id count, need {:?}, but got {:?}",
+                region_to_split.get_peers(),
+                req.get_new_peer_ids()
+            ));
+        }
+        keys.push_back(split_key.to_vec());
+    }
+
+    util::check_key_in_region(keys.back().unwrap(), region_to_split)?;
```
Comment (BusyJay) [inline components/raftstore/src/store/fsm/apply.rs:1933]: Seems `check_key_in_region_exclusive` makes more sense. Though both functions work the same here.

File: components/engine_traits/src/misc.rs [hunk 3]
```diff
@@ -95,4 +97,6 @@ pub trait MiscExt: CfNamesExt + FlowControlFactorsExt {
     ) -> Result<Option<(u64, u64)>>;
 
     fn is_stalled_or_stopped(&self) -> bool;
+
+    fn create_checkpoint(&self, basedb_out_dir: &Path) -> result::Result<(), String>;
```
[the remark this answers is not in the document]
Comment (SpadeA-Tang) [reply components/engine_traits/src/misc.rs:101]: Good suggestion.

File: components/raftstore/src/store/fsm/apply.rs (hunk 2 above)
Comment (SpadeA-Tang) [reply components/raftstore/src/store/fsm/apply.rs]: done

File: components/engine_traits/src/checkpoint.rs [hunk 4]
```diff
@@ -0,0 +1,9 @@
+// Copyright 2022 TiKV Project Authors. Licensed under Apache-2.0.
+
+use crate::Result;
+
+pub trait Checkpoint {
+    type Checkpointer;
```
Comment (BusyJay) [inline components/engine_traits/src/checkpoint.rs:6]: How to use `Checkpointer`?

File: components/engine_traits/src/checkpoint.rs [hunk 5]
```diff
@@ -0,0 +1,9 @@
+// Copyright 2022 TiKV Project Authors. Licensed under Apache-2.0.
+
+use crate::Result;
+
+pub trait Checkpoint {
```
Comment (BusyJay) [inline components/engine_traits/src/checkpoint.rs:5]: It should be named `Checkpointable` just like `Writeable`.

Comment (SpadeA-Tang) [reply components/engine_traits/src/checkpoint.rs]: Done

File: components/engine_traits/src/checkpoint.rs [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+pub trait Checkpointable {
+    type Checkpointer: Checkpointer;
+
+    fn new_checkpointer(&self) -> Result<Self::Checkpointer>;
+}
+
+pub trait Checkpointer {
+    fn create_at(
+        &mut self,
+        basedb_out_dir: &Path,
```
Comment (BusyJay) [inline components/engine_traits/src/checkpoint.rs:16]:
```suggestion
        db_out_dir: &Path,
```

**Review by BusyJay — APPROVED** [review]
Rest LGTM

File: components/engine_rocks/src/checkpoint.rs [hunk 7]
```diff
@@ -0,0 +1,55 @@
+// Copyright 2022 TiKV Project Authors. Licensed under Apache-2.0.
+
+use std::path::Path;
+
+use engine_traits::{Checkpointable, Checkpointer, Result};
+
+use crate::{r2e, RocksEngine};
+
+impl Checkpointable for RocksEngine {
```
Comment (tabokie) [inline components/engine_rocks/src/checkpoint.rs:9]: RocksDB interface is too cumbersome, the overhead of `new_checkpointer` is negligible. Let's just expose only one trait and one interface `create_checkpoint`.

Comment (SpadeA-Tang) [reply components/engine_rocks/src/checkpoint.rs:9]: How about your suggestions about this? @BusyJay

Comment (BusyJay) [reply components/engine_rocks/src/checkpoint.rs:9]: I prefer to keep align with rocksdb's API.

**Review by tabokie — APPROVED** [review]

**Pull request merged**